### ISSUE TITLE
Repo cleanup: extract utils, add CLAUDE.md and README, establish standards

### ISSUE THREAD

**Issue by ofekvs18** [ofekvs18/UnderGradProject#4]
## Goal

Clean up the repository structure, reduce duplication, add project documentation, and establish coding standards for all future method scripts.

## Current state

```
biomarker-pipeline/
├── .git/
├── .gitignore
├── data/                          # gitignored, user provides CSV
├── requirements.txt
├── results/                       # generated outputs
│   ├── sanity_check_results.csv
│   ├── sanity_summary.txt
│   ├── roc_curves.png
│   └── method1_threshold/
└── src/
    ├── checkpoint5_sanity.py      # 150 lines
    ├── method1_threshold.py       # 230 lines
    └── add_pr_metrics.py          # new, not yet run
```

### Problems
1. **Duplicated code** across scripts: data loading, train/test splitting, metric computation (precision/recall/F1 from tp/fp/fn), ROC curve plotting, results saving
2. **No README** — someone cloning the repo has no idea what this is or how to run it
3. **No CLAUDE.md** — Claude Code doesn't know the project conventions
4. **Inconsistent naming** — scripts are named by checkpoint/method number which will get confusing as more are added
5. **Hardcoded paths** — `DATA_PATH = "data/ra_modeling_data.csv"` repeated in every file
6. **No shared configuration** — feature lists, baseline AUC, thresholds all hardcoded in each script

### 1. Create `src/utils.py` — shared utilities

Extract the following repeated code into reusable functions:

```python
# src/utils.py

import pandas as pd
import numpy as np
from pathlib import Path

# ── Paths ────────────────────────────────────────────────────────────────
DATA_DIR = Path("data")
RESULTS_DIR = Path("results")
DATA_PATH = DATA_DIR / "ra_modeling_data.csv"

# ── Constants ────────────────────────────────────────────────────────────
META_COLS = {"subject_id", "is_case", "split"}
BASELINE_AUC = 0.658  # all-features logistic regression

# ── Data loading ─────────────────────────────────────────────────────────
def load_data(path=DATA_PATH):
    """Load the modeling CSV and return (df, feature_names)."""
    df = pd.read_csv(path)
    features = [c for c in df.columns if c not in META_COLS]
    return df, features

def get_splits(df):
    """Return (train_df, test_df) using the pre-computed split column."""
    return df[df["split"] == "train"].copy(), df[df["split"] == "test"].copy()

# ── Metrics ──────────────────────────────────────────────────────────────
def compute_binary_metrics(y_true, y_pred):
    """Compute precision, recall, F1, F2 from binary predictions."""
    tp = int(((y_pred == 1) & (y_true == 1)).sum())
    fp = int(((y_pred == 1) & (y_true == 0)).sum())
    fn = int(((y_pred == 0) & (y_true == 1)).sum())
    precision = tp / (tp + fp) if (tp + fp) > 0 else 0.0
    recall = tp / (tp + fn) if (tp + fn) > 0 else 0.0
    f1 = 2 * precision * recall / (precision + recall) if (precision + recall) > 0 else 0.0
    f2 = 5 * precision * recall / (4 * precision + recall) if (4 * precision + recall) > 0 else 0.0
    return {"precision": precision, "recall": recall, "f1": f1, "f2": f2, "tp": tp, "fp": fp, "fn": fn}

def find_youden_threshold(y_true, scores):
    """Find optimal threshold via Youden's index on ROC curve. Returns (threshold, fpr, tpr)."""
    from sklearn.metrics import roc_curve
    fpr, tpr, thresholds = roc_curve(y_true, scores)
    j = tpr - fpr
    best_idx = int(np.argmax(j))
    return float(thresholds[best_idx]), fpr, tpr

# ── Output helpers ───────────────────────────────────────────────────────
def ensure_dir(path):
    """Create directory if it doesn't exist."""
    Path(path).mkdir(parents=True, exist_ok=True)
```

### 2. Refactor existing scripts to use `src/utils.py`

- `checkpoint5_sanity.py` → import from utils, remove duplicated metric code
- `method1_threshold.py` → import from utils, remove duplicated metric code
- `add_pr_metrics.py` → import from utils
- Verify all scripts still produce the same output after refactoring

### 3. Create `CLAUDE.md` in project root

```markdown
# Biomarker Pipeline — Project Guide for Claude Code

## What this project does
Biomarker discovery for Rheumatoid Arthritis from CBC (Complete Blood Count) data.
Uses MIMIC-IV dataset. Compares multiple methods for generating biomarker formulas.

## How to run
All scripts run from the project root: `python src/<script_name>.py`
Data must be in `data/ra_modeling_data.csv` (not tracked in git).

## Coding standards
- **One file per method** in `src/`, named `method<N>_<name>.py`
- **Shared code goes in `src/utils.py`** — never duplicate data loading or metric computation
- **No classes, no frameworks** — keep it flat, functions only
- **Use the pre-computed split column** — NEVER re-split the data
- **All results go to `results/`** — each method gets its own subdirectory
- **Report these metrics for every method**: AUC-ROC, AUC-PR, precision, recall, F1, F2, precision@recall(0.25, 0.50, 0.75)
- **Print progress** to stdout as scripts run
- **Use `src/utils.py` constants** for paths, baseline AUC, feature lists

## Data
- `data/ra_modeling_data.csv` — exported from BigQuery, gitignored
- Columns: subject_id, is_case, split, hct, hgb, mch, mchc, mcv, plt, rbc, rdw, wbc
- Split is immutable: train (80%) / test (20%), patient-level, deterministic
- ~1% positive rate (RA cases) — always use imbalanced-aware metrics

## Key numbers
- Train: 1,273 cases / 115,955 controls
- Test: 300 cases / 28,976 controls
- Baseline AUC-ROC: 0.658 (all-features logistic regression)

## File structure
```
biomarker-pipeline/
├── CLAUDE.md              # this file
├── README.md              # project overview
├── requirements.txt
├── data/                  # gitignored
├── results/               # generated outputs, mostly gitignored
│   └── experiment_log.md  # tracked — manual experiment tracking
└── src/
    ├── utils.py           # shared utilities (data loading, metrics, paths)
    ├── checkpoint5_sanity.py
    ├── method1_threshold.py
    ├── add_pr_metrics.py
    └── (future method scripts)
```
```

### 4. Create `README.md` in project root

Should include:
- Project title and one-line description
- Research question: "Can formulas derived from routine CBC values predict Rheumatoid Arthritis?"
- Quick start: how to set up and run
- Data: where it comes from (MIMIC-IV, BigQuery), how to get it (link to SQL queries)
- Methods overview: threshold optimization → random formulas → GP → LLM (with status: done/in-progress/planned)
- Results: point to `results/experiment_log.md`
- Author and advisor info

### 5. Update `.gitignore`

Add:
```
# Don't ignore experiment log (it's our tracking doc)
!results/experiment_log.md
```

## Order of operations

1. Create `src/utils.py` first
2. Refactor `checkpoint5_sanity.py` to use it — verify same output
3. Refactor `method1_threshold.py` to use it — verify same output
4. Update `add_pr_metrics.py` to use it
5. Create `CLAUDE.md`
6. Create `README.md`
7. Update `.gitignore`
8. Commit everything as "chore: restructure repo, add docs, extract shared utils"

## What NOT to do

- Do NOT rename the scripts yet — the current names are fine for now
- Do NOT change any logic or results — this is purely structural
- Do NOT add new dependencies
- Do NOT create abstract base classes or a plugin system — keep it simple
